### PR TITLE
Add Cookie based authentication for micro-gateway 2.X branch

### PR DESCRIPTION
**Pull request by Chamindu36**
## Purpose
Provide authentication for product micro-gateway with basic authentication which uses usernames and passwords for authentication.

## Goals
Providing feature of enabling Cookie based authentication for to product micro-gateway along with OAuth2 token based authentication. User will be benefited with using only OAuth2 token based authentication alone, or using Cookie to authenticate micro-gateways.

## Approach
When an user invoke an API with a cookie, micro-gateway has to validate that cookie prior to the response. The list of cookies included in the HTTP request which use to authenticate, have to be extracted from the request. From all extracted cookies,their respective session ID value has to be extracted properly.
         The Authn filter will check incoming request to micro-gateway and determine whether it contains header as "Authorization" or header as "Cookie". If header is equals to "Cookie" then the cookie validation process will be executed and cookie will be validated. If not it will execute as a normal request which contains header as "Authorization". The session ID of the required cookie can be provided to server as a direct key value pair at the micro-gateway server startup.

https://docs.google.com/document/d/1HospX0LmjwRAuFSUfFXY9uGchP3xjM5qFFYJvImkFQ0/edit?usp=sharing

## User stories

User Story | Micro-gateway that is only secured with oauth2.
-- | --
Overview | As an user, I need to secure my API only with oauth2.
Persona | API creator, API Publisher
Prerequisites | None
Narrative | User setup micro-gateway.Then user build micro-gateway and no additional steps are needed.
Acceptance Criteria | Micro-gateway that is secured with oauth2 token should be allowed to access, only if token validation passesMicro-gateway secured with this Oauth2 cannot be validated usernames and passwords.

User Story | Micro-gateway that is secured with oauth2 or with cookie.
-- | --
Overview | As an user, I need to secure my API only with oauth2 and give support for cookie authentication.
Persona | API creator, API Publisher
Prerequisites | Oauth2 token of Authorized cookie.
Narrative | User setup micro-gateway.Then user build micro-gateway and no additional steps are needed.Provide Valid cookie with Cookie keyword as an environment variable at server startup.
Acceptance Criteria | Micro-gateway that is secured with oauth2 token should be allowed to access, only if token validation passes.Micro-gateway secured with this Oauth2 can be passed if the request contains cookie with valid Oauth2 token.

## Documentation
Doc has Impact

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   CookieAuthTestCase

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import ballerina/reflect;
+
+
+@Description { value: "Representation of the Cookie Based authetication logic" }
+public type CookieBasedAuth object {
+
+    @Param { value: "request: Request instance" }
+    @Return { value: "FilterResult:Validation of cookies in request" }
+    public function processRequest( http:Request request) returns string|error {
+        boolean isAuthorized;
```
Comment (pubudu538) [inline components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal]: What is the usage of this boolean?

File: components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            }
+        }
+        error notFound = { message: "No Authorized cookie is found" };
+        return notFound;
+    }
+
+    @Param { value: "request: Request instance" }
+    @Return { value: "FilterResult:Check whether Validation of cookies in request is authenticated" }
+    public function isCookieAuthed( http:Request request) returns boolean {
+        boolean isAuthorized;
```
Comment (pubudu538) [inline components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal]: What is the usage of this boolean?

File: components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal [hunk 3]
```diff
@@ -71,14 +71,14 @@ public type AuthnFilter object {
         //Create auth handler chain with providerIds in service file
         http:AuthHandlerRegistry registry;
         http:AuthProvider[] authProviders = getAuthProviders();
-        foreach i in authProvidersIds{
-            if (i == "oauth2"){
+        foreach authProvidersId in authProvidersIds{
+            if (authProvidersId == "oauth2"){
```
Comment (pubudu538) [inline components/micro-gateway-core/src/main/ballerina/gateway/filters/authn_filter.bal]: Let's move oauth2 to a constant.

File: components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal [hunk 4]
```diff
@@ -100,8 +112,13 @@ public function getAuthenticationFailureMessage(int errorCode) returns string {
         errorMessage = INVALID_SCOPE_MESSAGE;
     } else if (errorCode == API_AUTH_INVALID_BASICAUTH_CREDENTIALS) {
         errorMessage = API_AUTH_INVALID_BASICAUTH_CREDENTIALS_MESSAGE;
+
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/constants/api_security_constants.bal]: Let's remove this extra line.

File: components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    public function processRequest( http:Request request) returns string|error {
+        boolean isAuthorized;
+
+        //get required cookie as config value
+        string requiredCookie = config:getAsString(COOKIE_HEADER, default = "");
+
+        //extract cookies from the incoming request
+        string authHead = request.getHeader(COOKIE_HEADER);
+        string[] cookies = authHead.trim().split(";");
+        foreach cookie in cookies{
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal]: Minor formatting issue

File: components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    public function isCookieAuthed( http:Request request) returns boolean {
+        boolean isAuthorized;
+
+        //get required cookie as config value
+        string requiredCookie = config:getAsString(COOKIE_HEADER, default = "");
+
+        //extract cookies from the incoming request
+        string authHead = request.getHeader(COOKIE_HEADER);
+        string[] cookies = authHead.trim().split(";");
+        foreach cookie in cookies{
```
Comment (tharikaGitHub) [inline components/micro-gateway-core/src/main/ballerina/gateway/cookie.bal]: Minor formatting issue

File: tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/CookieAuthTestCase.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (tharikaGitHub) [inline tests/src/test/java/org/wso2/micro/gateway/tests/toolkit/CookieAuthTestCase.java]: Let's update the license year

**Pull request merged**